add lighthouse-user-flows doc (#6660)
GoogleChrome/web.dev: src/site/content/en/blog/lighthouse-user-flows/index.md

---
title: 'Lighthouse user flows'
subhead: Try out a new Lighthouse API to measure throughout a user flow.

> description: Try out a new Lighthouse API to measure throughout a user flow.

Reviewer: Is there a word missing in these? "To measure [potentially missed word] through a user flow?"

authors:
  - bckenny
date: 2021-11-03
hero: image/MtjnObpuceYe3ijODN3a79WrxLU2/NbfbZ81Sx2ZlFvRwTiBV.png
alt: A Lighthouse flow report summary
tags:
  - blog # blog is a required tag for the article to show up in the blog.
  - performance
  - web-vitals
  - lighthouse
---

> Lighthouse is a fantastic tool for testing performance and best practices during initial page load. However, it's traditionally been difficult to use Lighthouse to analyze other aspects of the life of a page. Page loads with a warm cache, or an activated Service Worker, or during user interactions are all scenarios that have been difficult to test using Lighthouse.

Reviewer: ```suggestion
Lighthouse is a fantastic tool for testing performance and best practices during initial page load. However, it's traditionally been difficult to use Lighthouse to analyze other aspects of the life of a page, such as:

- Page loads with a warm cache
- Pages with an activated Service Worker
- Accounting for potential user interactions
```

{% endAside %}

## Navigations

The new Lighthouse "navigation" mode is actually giving a name for the (up until now) standard Lighthouse behavior: analyze the cold load of a page. This is the mode to use to monitor page load performance, but user flows also open up the possibility of new insights.

To script Lighthouse capturing a page load:

- use puppeteer to open the browser
- start a Lighthouse user flow
- navigate to the target URL

```js
import fs from 'fs';
import open from 'open';
import puppeteer from 'puppeteer';
import {startFlow} from 'lighthouse/lighthouse-core/fraggle-rock/api.js';

async function captureReport() {
  const browser = await puppeteer.launch({headless: false});
  const page = await browser.newPage();

  const flow = await startFlow(page, {name: 'Single Navigation'});
  await flow.navigate('https://web.dev/performance-scoring/');

  await browser.close();

  const report = flow.generateReport();
  fs.writeFileSync('flow.report.html', report);
  open('flow.report.html', {wait: false});
}

captureReport();
```

> This is the simplest "flow". When opened, the report shows a summary view with only the single step; clicking on that step will reveal a traditional Lighthouse report for that navigation.

Author: > Not sure "flow" needs to be in quotes in this case.

ha, almost seems to be mocking our own name for it :)

> This is the simplest "flow". When opened, the report shows a summary view with only the single step; clicking on that step will reveal a traditional Lighthouse report for that navigation.

Reviewer: I've totally done this, too!
